### PR TITLE
Replace renderer with WebGL renderer

### LINKED ISSUE #25: "Touching sprite" block behavior with clones?
In Scratch, the "touching (sprite)" block will return true if the current sprite is touching the other sprite or any of its clones. In scratch-js, it only returns true if it's touching the original sprite.

I think a good solution would be to retain the current behavior if you pass a sprite *instance* to `touching()`, but to use Scratch's behavior if you pass the sprite's *base class*. I'd like to get your opinion on it.

### LINKED ISSUE #27: Rotation style API?
Scratch allows you to set a sprite's rotation style to "all around", "left-right", and "don't rotate".

I can think of two ways to implement this in scratch-js. One is to use string keys:

```js
this.rotationStyle = "all around";
this.rotationStyle = "left-right";
this.rotationStyle = "don't rotate"
```

The other is to use an enum:
```js
// these don't have to be Symbols; the important thing is they're stored in an enum
Sprite.RotationStyle = {
  AllAround: Symbol("all around"),
  LeftRight: Symbol("left-right"),
  DontRotate: Symbol("don't rotate")
}
```

which could be used like so:
```js
this.rotationStyle = Sprite.RotationStyle.AllAround;
this.rotationStyle = Sprite.RotationStyle.LeftRight;
this.rotationStyle = Sprite.RotationStyle.DontRotate;
```

I think the second option is less error-prone, but it would be the first time a static property is added to the `Sprite` class.

### PR DESCRIPTION
**Pull request by adroitwhiz**
Resolves #25 
Resolves #27

I've finished\* the new renderer, and it's ready for review!

## Changes to Existing APIs
- `Renderer.checkPointCollision` now takes coordinates in "Scratch-space" ((-240, 240) and (-180, 180), positive y-values are up), instead of "canvas-space" ((0, 480) and (0, 360), positive y-values are down).
- `Renderer.getBoundingBox` now returns bounds in "Scratch-space" instead of "canvas-space".
- `Input` now takes a `Stage` object and a `<canvas>` in its constructor, as opposed to only a `<canvas>`.
- `Renderer` now takes a `Project` and optional render target in its constructor, as opposed to an optional render target and `{w, h}` object.
- `Renderer.checkSpriteCollision` can now take either a sprite or an `Array` of sprites as its second argument. In addition, it has a fourth argument, `sprColor`, which allows you to mask the collision check to a certain color of the source sprite. This can be used for "color is touching sprite".
- As far as I can tell, `Sprite.normalizeScratch` does the same thing as `Sprite.normalizeDeg` when operating correctly, so I've removed it.
- `Sprite.normalizeDeg` has been fixed to work properly with negative directions.
- `Sprite.touching` can now take a `Color` as its parameter. This works like the "touching color" block in Scratch.
- Speech bubbles now flip when they hit the right edge of the stage, rather than horizontally shrinking.

## New APIs
- You can now pass a `width` and `height` into the `Stage` constructor's first parameter. This specifies the `Stage`'s "Scratch-space" size, and can be used to e.g. create a widescreen stage.
- You can now resize a `Project`'s render target with CSS, and the canvas will automatically be resized to fit.
- Sprites now have the property `effects`, implemented internally as `_EffectMap`. This can be used to set or clear the sprite's graphic effects, e.g:
    - ```js
      this.effects.clear();
      this.effects.color = 50;
      this.effects.ghost += 5;
      ```
- Sprites now have the property `rotationStyle`. The `Sprite` class also now has the static enum property `RotationStyle`. This can be used to set the sprite's rotation style, e.g:
  - ```js
    this.rotationStyle = Sprite.RotationStyle.LEFT_RIGHT;
    ```
- Sprites now have the method `colorTouching`. When passed a `Color`, it works like the "color is touching color" block in Scratch. When passed a `Sprite` or list of `Sprite`s, it can operate as "color is touching sprite(s)", a feature unique to `scratch-js`.

## \*Not-yet-implemented APIs
There are certain things still in the #23 checklist that I haven't included in this PR:
- Layer ordering
  - This would involve restructuring non-renderer-related APIs (mostly `Project`).
- Convex hull generation / "if on edge, bounce"
  - This would involve storing some renderer-related state in `Sprite` (namely, the convex hull). I'll save that for a later PR.
- Subpixel positioning of SVG viewboxes
  - This would involve rethinking the way costume images (and potentially projects) are loaded, and would involve placing some renderer-related code in `Costume`.

## Future work
- Move sprites' transform matrices into the `Sprite` instances themselves
  - This also requires dipping into the `Costume` loading code (we need to make sure we recalculate the sprite's transform matrix when its costume's image is loaded and the `width` and `height` are available), which once again leads into the deeper discussion of how to handle `Costume`s whose images have not yet finished loading.
- Add a canvas-based renderer back in
  - The existing canvas-based renderer would require some modification, but not too much.

### CODE REVIEW

File: scratch-js/Sprite.mjs [hunk 1]
```diff
@@ -365,11 +417,35 @@ export class Sprite extends SpriteBase {
...
+      return this._project.renderer.checkColorCollision(this, target);
     }
 
     return this._project.renderer.checkSpriteCollision(this, target, fast);
   }
 
+  colorTouching(color, target) {
+    if (typeof target === "string") {
+      console.error(
+        `Cannot find target "${target}" in "touching". Did you mean to pass a sprite class instead?`
```
Comment (PullJosh) [inline scratch-js/Sprite.mjs]: We should probably give a similar error message if `color` is not an instance of `Color`

File: scratch-js/renderer/Shaders.mjs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+#if defined(EFFECT_whirl) || defined(EFFECT_fisheye)
+const vec2 CENTER = vec2(0.5, 0.5);
+#endif
+
+#ifdef DRAW_MODE_COLOR_MASK
+uniform vec4 u_colorMask;
+
+// TODO: Scratch 2.0 and Scratch 3.0's CPU path check if the top 6 bits match,
+// which a tolerance of 3/255 should be equivalent to,
+// but Scratch's GPU path has a tolerance of 2/255.
```
Comment (PullJosh) [inline scratch-js/renderer/Shaders.mjs]: That's actually super interesting! This PR is teaching me a bunch of cool stuff. 😛

**Comment by adroitwhiz** [comment]
I'm 95% sure the only API changes are internal.

**Comment by PullJosh** [comment]
@adroitwhiz This seems like an appropriate time to add a `contributors` field to `package.json`, if you want to be added. What name/url/email (if any) do you want included?

**Comment by adroitwhiz** [comment]
`adroitwhiz` is fine for my name, you can link to my [GitHub page](https://github.com/adroitwhiz), and my email is `adroitwhiz@protonmail.com`.

Thanks for including me as a contributor!

**Pull request merged**